### DISCUSSION TITLE
Why oh why?! "pybricksdev run" hangs?

### DISCUSSION THREAD

**Discussion by vascolp** [pybricks/support#1142] [Q&A]
I am having a lot of pybricksdev hanging when trying to run a program. 

It hangs almost always at 300b (`300/1.55k [00:30<02:06, 9.92B/s]`), no matter the size of the program, but with a bigger programs it's impossible, it always hangs...

- Command line: `pybricksdev.exe run ble -n "THUB 5" ligapc_teste.py`
- pybricksdev v1.0.0a46
- TechnicHub with  Pybricks 3.3.0b6 
- Cygwin bash on windows 10

pybricksdev gets stuck, or enters a very long timeout, I have to kill it...
Any ideas?

**Comment by vascolp** [comment]
Maybe if I stop using it for some time (like coming here to complain...) it works better, I was just able to run a big program. Is there something on windows that might confuse BLE buffers or whatever? just guessing...

**Reply by dlech** [reply]
Is the battery low?

Are there other Bluetooth devices that could be interfering?

**Comment by vascolp** [comment]
Battery level is OK according with pybricks.code.... 
Other devices... in terms of bluetooth devices I am in the middle of nowhere... So two phones, a TV set and a computer... Never had problems with them...
This is not the first time I have this, and I usually check battery level...

**Comment by vascolp** [comment]
Just noticed:
If I run pybricksdev with the hub disconnected, it timesout (long traceback...). Then, in the next run, it works fine.

**Comment by dlech** [comment]
There are more tips and instructions on how to log Bluetooth packets at https://github.com/orgs/pybricks/discussions/270.

**Comment by vascolp** [comment]
The problem is not only with pybricksdev, but also in pybricks.code, athough pybricks.code is a bit better.
I have a Microsoft sculpt keyboard and mouse. Turning it off seems to make things a bit better...
But I have it for almost 3 years and, in with previous laptop, never had problems with pybricks... It was a laptop from 2016/17, so with an older bluetooth.

If I get a bluetoorh dongle like [this](https://github.com/pybricks/support/issues/921#issuecomment-1404965804) do you think it might work better?

**Reply by dlech** [reply]
It is hard to say. Many users have reported good results with the Cambridge Silicone Radio chipset. But there could also be other hardware issues with your laptop.

**Comment by vascolp** [comment]
Why pybricksdev fails to me many times while transfering a program to run it but it never fails while flashing a firmware?
Are flashing comunications more robust?

**Reply by laurensvalk** [reply]
Please use the instructions to log Bluetooth data as linked above. That would be the only way to find out.

It would be great if you want to do so both with Pybricksdev and Pybricks Code.

**Reply by vascolp** [reply]
Well with pybricksdev -d it fails always in the same place:
There is a pause between the 3% line and the following lines, like if there was a timeout. Then it gets stucked for ever in bleak disconnection.
Meanwhile, the hub disconnects and goes to advertising mode.

```
2023-07-07 10:20:00,689: INFO: pybricksdev.connections.pybricks: Connected successfully!
2023-07-07 10:20:00,808: DEBUG: bleak.backends.winrt.client: Read Characteristic 000A : bytearray(b'3.3.0b7')
2023-07-07 10:20:00,924: DEBUG: bleak.backends.winrt.client: Read Characteristic 000C : bytearray(b'1.3.0')
2023-07-07 10:20:01,047: DEBUG: bleak.backends.winrt.client: Read Characteristic 000E : bytearray(b'\x01\x97\x03\x80\x00\x00\x00')
2023-07-07 10:20:01,166: DEBUG: bleak.backends.winrt.client: Read Characteristic 0014 : bytearray(b'\x9b\x00\x03\x00\x00\x00t?\x00\x00')
2023-07-07 10:20:01,440: DEBUG: pybricksdev.compile: missing modules: ['micropython', 'pybricks', 'pybricks.hubs', 'pybricks.iodevices', 'pybricks.parameters', 'pybricks.pupdevices', 'pybricks.tools', 'uerrno', 'uselect', 'usys']
2023-07-07 10:20:31,847: DEBUG: bleak.backends.winrt.client: max_pdu_size_changed_handler: 23
  3%|██▊                                                                                                           | 300/12.0k [00:30<19:39, 9.92B/s]
2023-07-07 10:20:31,854: INFO: pybricksdev.connections.pybricks: Disconnecting...
2023-07-07 10:20:31,853: DEBUG: bleak.backends.winrt.client: session_status_changed_event_handler: id: BluetoothLE#BluetoothLE80:b6:55:38:b1:9c-32:65:
65:ef:cd:9a, error: BluetoothError.SUCCESS, status: GattSessionStatus.CLOSED
2023-07-07 10:20:31,855: DEBUG: bleak.backends.winrt.client: Disconnecting from BLE device...

```

**Reply by vascolp** [reply]
Using Wireshark/BTP I get the following log (a pcapng file), hope it helps, it´s the first time I use this...:
~
[pybricksdev_run_stuck.zip](https://github.com/pybricks/support/files/11980807/pybricksdev_run_stuck.zip)

**Reply by vascolp** [reply]
And here is a file with Pybricks Code:
[pybricks_code_run_stuck.zip](https://github.com/pybricks/support/files/11981990/pybricks_code_run_stuck.zip)

**Reply by dlech** [reply]
Thanks for the logs. The hub isn't responding to a write request, so it looks like a bug in the firmware.

**Comment by ivorjawa** [comment]
I’ve seen this problem, it’s in PyBricks dev/ble/__init__.py in match_uuid_and_name.

if adv.local_name is None:
    # have not received SCAN_RSP yet
    return False

my fix ONLY FOR DEVICES WITH ASSIGNED NAMES:

above the above if:

if adv.platform_data[0].name() == name:
    return True
